### PR TITLE
Route all LLM calls through OpenClaw gateway — provider-agnostic model fallback chain (#87)

### LINKED ISSUE #87: Route all LLM calls through OpenClaw gateway — provider-agnostic model fallback chain
## Problem

The plugin currently makes **direct API calls** to LLM providers:
- **Gemini**: Direct REST `fetch()` to `generativelanguage.googleapis.com` (in `chat.ts`)
- **OpenAI/Claude**: Via the `openai` SDK client, but with hardcoded model names

This means:
1. Users **must** have a Gemini, OpenAI, or Claude API key — no other provider works
2. The plugin **bypasses** OpenClaw gateway's model routing, fallback chains, and provider abstraction
3. Users who only have e.g. a Kiwi, Microsoft Copilot, Groq, or local Ollama key cannot use the plugin's LLM features (distillation, reflection, self-correction, proposals, etc.)
4. No benefit from gateway features like rate-limit handling, key rotation, or cost tracking

### 1. Route all LLM calls through the OpenClaw gateway OpenAI-compatible endpoint

Replace direct Gemini REST calls and hardcoded OpenAI SDK usage with calls through `ctx.openai` (the gateway-proxied client). The gateway already handles model resolution and provider routing.

### 2. Configurable ordered model preference list

```yaml
plugins:
  entries:
    openclaw-hybrid-memory:
      config:
        llm:
          # Ordered preference — first available/working model wins
          default:
            - gemini-2.0-flash
            - claude-sonnet-4
            - gpt-4o-mini
          heavy:
            - gemini-2.0-flash-thinking
            - claude-opus-4
            - gpt-4o
          # Final fallback: use the user's default model from openclaw.yaml
          fallbackToDefault: true
```

### 3. Fallback chain behavior

For each LLM call:
1. Try models from the preference list in order
2. If a model fails (no key, 401, 403, provider unavailable), try the next
3. If all preferred models fail and `fallbackToDefault: true`, use the user's configured default model
4. Only fail if absolutely no model can handle the request

### 4. Remove provider-specific code paths

- Delete the direct Gemini `fetch()` path in `chat.ts`
- Delete `isGeminiModel()` / `resolveGeminiApiKey()` helpers
- Remove `geminiApiKey` threading through every function
- Single code path: `ctx.openai.chat.completions.create({ model, ... })`

### 5. Embeddings

Same pattern for embeddings — route through gateway if possible, with a preference list.

## Impact

- **Any provider works**: Kiwi, Copilot, Groq, Ollama, OpenRouter, etc.
- **Zero config for basic users**: `fallbackToDefault: true` means it just works with whatever the user has
- **Power users can optimize**: Specify exact model preferences per task type
- **Simpler codebase**: One LLM call path instead of three
- **Gateway benefits**: Rate limiting, key rotation, cost tracking, failover — all for free

## Files Affected

- `services/chat.ts` — Main rewrite (remove Gemini path, single gateway path)
- `config.ts` — New `llm` config schema, remove `getDefaultCronModel` provider detection
- `lifecycle/hooks.ts` — Remove `geminiApiKey` plumbing
- `cli/handlers.ts` — Simplify model resolution for cron jobs
- All consumers of `chatComplete` / `chatCompleteWithRetry`

## Migration

- Existing `distill.apiKey` (Gemini) and `claude.apiKey` configs become optional/deprecated
- Users with those keys keep working (gateway routes to the right provider)
- New users just need any working provider in their OpenClaw config

### PR DESCRIPTION
**Pull request by markus-lassfolk**
## Summary

Implements [issue #87](https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/87): route all LLM calls through the OpenClaw gateway with a configurable model preference list and fallback chain.

### Core implementation
- **Single code path:** All chat/completion goes through `ctx.openai.chat.completions.create()`; removed direct Gemini REST and provider-specific branches.
- **`llm` config:** `default` and `heavy` ordered model lists, `fallbackToDefault`, `fallbackModel`. `getLLMModelPreference(cfg, tier)` returns the list to try; `getDefaultCronModel` uses first model when `llm` is set, else legacy provider-based resolution.
- **Removed:** `isGeminiModel`, `isAnthropicModel`, `resolveGeminiApiKey`, `geminiApiKey` parameter from chat/reflection/tools/handlers.
- **Embeddings:** Use same gateway client (`Embeddings(openai, model)` in init-databases).
- **Sentry/GlitchTip:** `capturePluginError` in `chatComplete`, `withLLMRetry`, `chatCompleteWithRetry`, and `embed` catch blocks.

### Docs
- **New:** [docs/LLM-AND-PROVIDERS.md](docs/LLM-AND-PROVIDERS.md) — prerequisites (embedding required, chat optional), how LLMs are used per feature, gateway routing, `llm` config, legacy behaviour.
- **Updated:** README, QUICKSTART, CONFIGURATION, OPERATIONS, CLI-REFERENCE, SESSION-DISTILLATION, index with gateway + `llm` and link to LLM-AND-PROVIDERS.

### Tests
- Chat tests: removed Gemini REST/fetch and `isGeminiModel`; added gateway-path tests for `gemini-2.0-flash`.
- Config tests: added `llm` parsing and `getLLMModelPreference` / `getDefaultCronModel` behaviour.

## Config example

```yaml
plugins:
  entries:
    openclaw-hybrid-memory:
      config:
        llm:
          default:
            - gemini-2.0-flash
            - claude-sonnet-4
            - gpt-4o-mini
          heavy:
            - gemini-2.0-flash-thinking
            - claude-opus-4
            - gpt-4o
          fallbackToDefault: true
```

Closes #87.

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Medium Risk**
> Touches core LLM routing/model selection used across multiple maintenance jobs and interactive commands, so misconfiguration or gateway issues could break distill/reflect/self-correction workflows. Changes are mostly additive with legacy fallbacks and expanded tests/docs, reducing rollout risk.
> 
> **Overview**
> **All chat/completion calls are now gateway-routed and provider-agnostic.** The plugin removes direct Gemini REST/provider-specific routing and instead always uses the gateway’s OpenAI-compatible client, with stronger error capture in `chatComplete`/retry/fallback paths.
> 
> Adds optional `llm` config (`default`/`heavy` ordered model lists + fallback behavior) and updates cron/CLI features (distill, ingest, reflection, self-correction, proposals, HyDE) to pick models from these tiers with fallbacks; legacy `distill`-based selection remains when `llm` is unset. Embeddings gain optional `embedding.models` fallback (same-dimension enforced) and are instantiated using the gateway-proxied OpenAI client.
> 
> Hardens persona proposals by scoping proposal generation via `autoRecall.scopeFilter` warning, capping confidence for whole-file replacements (esp. `SOUL.md`), and allowing apply in non-git workspaces (only auto-commit when inside a git repo); adds tests and updates docs/new `LLM-AND-PROVIDERS.md` to reflect the new config and prerequisites.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2034196bcdd1874e0c7131d7e84bf061f0aedf9e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by markus-lassfolk** [comment]
## PR #88 — All council + bot feedback addressed

All blocking and should-fix issues from the 4-model council review + Copilot threads are now resolved across two commits on this branch:

### Commit `71f4d95` (Council fixes — fallback, single-tier llm, proposals)
- **Fix 1+2 (Copilot threads — now resolved):** Removed hardcoded `gpt-4o`/`gpt-4o-mini` from `getLLMModelPreference`. Only appends `fallbackModel` when explicitly configured and not already in the list. Test updated to expect deduped output.
- **Fix 3:** `runGenerateProposalsForCli` now falls back to `cfg.distill?.fallbackModels ?? []` when preference list is single-model — matching distill/ingest/reflection handlers.
- **Fix 4:** Schema parser uses `||` (not `&&`) so `llm` is enabled with only `default` or only `heavy` configured (single-tier).
- **Fix 6 (partial):** Removed duplicate `proposals` registration from `register-tools.ts`; `manage.ts` is the single authoritative owner.
- **Fix 7:** Stale `getCronModelAlias` comments in `handlers.ts` updated to reference `getLLMModelPreference` / `getDefaultCronModel`.

### Commit `c00e780` (Remaining council + tsc fixes)
- **Fix 5:** `init-databases.ts` now passes `api.gatewayBaseUrl` as `baseURL` when available, so the OpenAI client routes through the gateway rather than bypassing it.
- **Fix 6 (complete):** Removed the dead `registerProposalsCli` import from `index.ts` (was imported but never called).
- **tsc clean:** Fixed pre-existing `opts.fallbackModels` type error in `cli-context.ts` (opts type doesn't include that property).
- **tsc clean:** Fixed pre-existing `mockResolvedValue` type error in `chat.test.ts` via `vi.mocked()`.

### Verification
- `npx tsc --noEmit` — ✅ clean
- `npx vitest run` — ✅ 656 passed (2 known pre-existing failures in `version-command.test.ts` excluded)
- Both unresolved Copilot threads resolved ✅

**Review by markus-lassfolk — COMMENTED** [review]
### Some heavy-tier calls don't use the new fallback chain (chatComplete vs chatCompleteWithRetry)
**Severity:** Medium

Several call sites now pick the first model from getLLMModelPreference(...)[0] but then invoke chatComplete(...) without passing fallback models (and without using chatCompleteWithRetry). For example, self-correction analysis / TOOLS rewrite appear to call chatComplete directly.

Impact: if the first preferred model fails (missing key, 429/5xx, provider outage), those operations will fail immediately even though the PR introduces model preference + fallback.

Suggested fix:
- Prefer chatCompleteWithRetry everywhere you expect model fallback behavior, and pass fallbackModels derived from getLLMModelPreference (slice(1), plus any legacy cfg.distill.fallbackModels if you intend to keep them).
- Add at least one test exercising a failing primary model and a succeeding fallback for a representative command (distill/reflection/self-correction).}

**Review by markus-lassfolk — COMMENTED** [review]
### Legacy model resolution priority regression (reflection.model overrides distill/claude)
**Severity:** High

In ,  now returns  *before* checking  / .

That changes the documented legacy priority (Gemini → Claude → OpenAI) and can cause cron jobs / CLI defaults to ignore a configured Gemini/Claude setup if the user happens to have  set (common), e.g. distillation heavy tier could incorrectly run on .

**Suggested fix:** Move the  override back under the OpenAI/embedding branch (or at least after distill/claude checks), so distill/claude remain higher priority in legacy mode. Add a unit test covering:  set +  set → returns distill default model (not reflection.model).

**Review by markus-lassfolk — COMMENTED** [review]
### Missing import for capturePluginError in embeddings.ts\n**Severity:** High\n\nThe new embeddings fallback logic calls , but  does not import it. This will throw a ReferenceError / fail build. Add  (and ensure path is correct) near the top of the file.

**Review by markus-lassfolk — COMMENTED** [review]
### Finding 2: `isLongContextModel` naming and potential for future expansion
**Severity:** Low

The `isLongContextModel` function in `services/chat.ts` checks for "gemini" or "thinking" in the model name to determine if it's a long-context model. This works for current Gemini models but ties the logic to specific keywords that might not generalize to all future long-context models from other providers. While it's functional for now, it's a hardcoded heuristic.

**Suggested Fix:** Consider if there's a more robust way to identify long-context models, perhaps by querying the gateway for model capabilities or having a configurable list. For now, it's acceptable, but it's a point of potential fragility.

**Review by markus-lassfolk — COMMENTED** [review]
### Finding 3: Missing test case for `getLLMModelPreference` with empty `llm` config
**Severity:** Low

The `config.test.ts` includes tests for `getLLMModelPreference` when `llm` is configured. However, there isn't an explicit test case for when `llm` is *not* present in the config, ensuring that it correctly falls back to `getDefaultCronModelLegacy` and returns the expected single model.

**Suggested Fix:** Add a test case to `config.test.ts` for `getLLMModelPreference` when `llm` is undefined or empty, verifying it correctly uses the legacy fallback logic.

**Review by markus-lassfolk — COMMENTED** [review]
### Missing import for capturePluginError in embeddings.ts
**Severity:** High

The new embeddings fallback logic calls `capturePluginError(...)`, but `services/embeddings.ts` does not import it. This will throw a ReferenceError / fail build. Add `import { capturePluginError } from "./error-reporter.js";` (and ensure path is correct) near the top of the file.

**Comment by markus-lassfolk** [comment]
## Flash (gemini-2.5-flash) Council Review

**Verdict:** REQUEST_CHANGES

**Findings:** 5
- Potential confusion in `llm.fallbackModel` documentation.
- `isLongContextModel` naming and potential for future expansion.
- Missing test case for `getLLMModelPreference` with empty `llm` config.
- Minor typo in `services/chat.ts` comment.
- `registerProposalsCli` moved without clear indication in diff.

**Overall Assessment:**

The PR makes significant and well-executed improvements, directly addressing the issues outlined in #87, #89, and #90.

*   **LLM Gateway Routing (#87):** The implementation of `llm` preference lists and gateway-routed LLM calls is a substantial upgrade, enhancing flexibility, maintainability, and leveraging OpenClaw gateway features. The `chat.ts` rewrite is clean and the `config.ts` changes correctly handle the new schema and legacy fallback.
*   **Proposal Generator Fixes (#89):** The prompt updates, scope filtering, and dynamic confidence adjustments for proposals are critical fixes that improve the reliability and safety of the persona proposal system.
*   **Optional Git Commit (#90):** The `isGitRepo` check and conditional commit logic correctly resolve the issue of proposals failing on non-Git workspaces.

The documentation updates are comprehensive and clearly explain the new configuration and behavior. The test suite also appears to be updated to reflect some of the new logic, particularly around `getLLMModelPreference`.

The identified findings are minor and do not detract from the overall quality and importance of this PR.

**Sign-off:** Flash (gemini-2.5-flash)

**Comment by markus-lassfolk** [comment]
## Codex (gpt-5.2-codex) Council Review

**Verdict:** REQUEST_CHANGES

**Findings:** 2
- Missing import for `capturePluginError` in `services/embeddings.ts` (build error)
- Legacy cron model priority changed: `reflection.model` now overrides distill/claude when it shouldn’t

**Overall assessment:** The gateway + llm routing work looks solid, but there’s a compile-time error and a legacy-model-selection regression that need fixing before merge.

File: extensions/memory-hybrid/cli/handlers.ts [hunk 1]
```diff
@@ -2028,14 +2030,16 @@ export async function runIngestFilesForCli(
     sink.log(`Processing batch ${b + 1}/${batches.length}...`);
     const userContent = ingestPrompt + "\n\n" + batches[b];
     try {
+      const ingestPref = getLLMModelPreference(getCronModelConfig(cfg), "default");
+      const ingestModel = model || ingestPref[0];
+      const ingestFallbacks = ingestPref.length > 1 ? ingestPref.slice(1) : cfg.distill?.fallbackModels;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/cli/handlers.ts]: Low priority / acknowledged. The preference list computation is lightweight (a few array operations on a small config object). The current structure keeps related logic co-located. Will address in a follow-up refactor if profiling shows it matters.

File: extensions/memory-hybrid/cli/handlers.ts [hunk 2]
```diff
@@ -2456,7 +2462,6 @@ export async function runSelfCorrectionRunForCli(
         temperature: 0.2,
         maxTokens: distillMaxOutputTokens(model),
         openai,
-        geminiApiKey: cfg.distill?.apiKey,
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/cli/handlers.ts]: Fixed in commit 4bde4ac (B3): `runSelfCorrectionRunForCli` now uses `chatCompleteWithRetry` for both the analyze and rewrite steps, with the full heavy-tier preference list as `fallbackModels`. The fallback bleed fix (S4) also ensures legacy `distill.fallbackModels` don't bleed through when `llm` is configured.

File: extensions/memory-hybrid/tools/utility-tools.ts [hunk 3]
```diff
@@ -232,7 +232,7 @@ export function registerUtilityTools(
             embeddings,
             openai,
             { defaultWindow: reflectionCfg.defaultWindow, minObservations: reflectionCfg.minObservations },
-            { window, dryRun: false, model: reflectionCfg.model, fallbackModels: cfg.distill?.fallbackModels, geminiApiKey: cfg.distill?.apiKey },
+            { window, dryRun: false, model: reflectionCfg.model, fallbackModels: cfg.distill?.fallbackModels },
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/tools/utility-tools.ts]: Already addressed in the PR: `utility-tools.ts` uses `getLLMModelPreference` and `getCronModelConfig` for all three reflection tools (memory_reflect, memory_reflect_rules, memory_reflect_meta). The S4 fix in 4bde4ac also prevents `cfg.distill?.fallbackModels` bleeding when `llm` is configured.

File: extensions/memory-hybrid/tests/config.test.ts [hunk 4]
```diff
@@ -532,6 +561,48 @@ describe("hybridConfigSchema.parse", () => {
...
+      ...validBase,
+      llm: {
+        default: ["gemini-2.0-flash", "gpt-4o-mini"],
+        heavy: ["gpt-4o"],
+        fallbackToDefault: true,
+        fallbackModel: "gpt-4o-mini",
+      },
+    });
+    const cronCfg = getCronModelConfig(cfg);
+    expect(getLLMModelPreference(cronCfg, "default")).toEqual(["gemini-2.0-flash", "gpt-4o-mini", "gpt-4o-mini"]);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/tests/config.test.ts]: The test at line 624 correctly reflects the deduped output after the dedupe guard was added. The test config has `heavy: ["gpt-4o"]` with `fallbackModel: "gpt-4o-mini"` — since gpt-4o-mini is not in the heavy list, it's appended once, giving `["gpt-4o", "gpt-4o-mini"]`. No duplicate.

File: extensions/memory-hybrid/setup/cli-context.ts [hunk 5]
```diff
@@ -625,6 +639,20 @@ export function registerCliWithHelp(
...
+        proposalsDb: handlerCtx.proposalsDb,
+        cfg: handlerCtx.cfg,
+        resolvedSqlitePath: handlerCtx.resolvedSqlitePath,
+        api,
+      });
+    } catch (err) {
+      capturePluginError(err instanceof Error ? err : new Error(String(err)), { subsystem: "registration", operation: "register-cli:proposals" });
+      throw err;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/setup/cli-context.ts]: Reviewed: `registerProposalsCli` is defined in `cli/proposals.ts` and is NOT called from `register-tools.ts` or any other file. The proposals CLI commands are registered only once in `cli/manage.ts` via `mem.command("proposals")`. No double registration occurs in the current code.

**Comment by markus-lassfolk** [comment]
## PR #88 — Final sweep complete

Commit `8286a2a` addresses the two remaining should-fix items:

### S3: Cross-user contamination without scopeFilter
**`cli/handlers.ts` — `runGenerateProposalsForCli`**

The previous warning only fired when `scopeKeys.size > 1`, which missed the common case where all facts carry `scope='global'` — a single scope key, but still potentially from multiple users. The warning now fires unconditionally whenever `autoRecall.scopeFilter` is not configured and facts are present, making the contamination risk visible every run.

### S8: Embedding model mismatch warnings
**`config.ts` — `parseHybridConfig`**

The catch block that silently dropped unknown `embedding.models` entries now calls `console.warn` naming the rejected model and suggesting valid alternatives. A second warn covers the case where all entries are recognized but have mismatched vector dimensions, which previously caused the entire list to be silently discarded.

### Bonus: narrower `ApplyProposalContext` type
`applyApprovedProposal` only uses `ctx.cfg.personaProposals.allowedFiles` and `ctx.api?.logger?.warn?.()`; the context type now reflects this. Structurally compatible with all existing callers.

**Verification:** `npx tsc --noEmit` clean · `npx vitest run` 673 passed (2 known pre-existing `version-command.test.ts` failures unchanged).

File: extensions/memory-hybrid/prompts/generate-proposals.txt [hunk 6]
```diff
@@ -1,12 +1,18 @@
...
+- Only propose changes for files listed in allowed_files. For each proposal, target exactly one file. Match insights to the correct file by purpose (see above).
+- You are given the current content of each identity file below. Use it. Prefer ADDING new sections or bullets rather than replacing. Only suggest replacing the entire file if the insight genuinely requires wiping existing content; in that case start suggestedChange with exactly: "Replace the entire file with the following:" and use confidence below 0.6 for such proposals.
+- suggestedChange: the exact markdown/text to add (e.g. a new section), or if truly necessary a full replacement starting with "Replace the entire file with the following:".
 - observation: 1-3 sentences explaining the evidence (which patterns/rules/meta support this) and why it belongs in the target file.
-- confidence: number between 0 and 1. Use 0.7+ for well-supported insights.
+- confidence: number between 0 and 1. Use 0.7+ for well-supported additions. Use 0.5 or lower for any proposal that would replace an entire file (especially SOUL.md — replacing safety/personality boundaries should never be high confidence).
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/prompts/generate-proposals.txt]: Fixed: Updated `generate-proposals.txt` — "below 0.6" → "0.6 maximum" to match `Math.min(confidence, 0.6)` in `capProposalConfidence`, and "0.5 or lower" → "0.5 maximum" for SOUL.md replacements. Both the prompt and the code now consistently express the same cap semantics.

File: extensions/memory-hybrid/config.ts [hunk 7]
```diff
@@ -634,8 +670,27 @@ export const hybridConfigSchema = {
...
+          vectorDimsForModel(m);
+          valid.push(m);
+        } catch {
+          /* skip unknown embedding model; consider logging when logger available (e.g. at init) so users see dropped entries */
+        }
+      }
+      if (valid.length > 0) {
+        const firstDim = vectorDimsForModel(valid[0]);
+        if (valid.every((m) => vectorDimsForModel(m) === firstDim)) {
+          embeddingModels = valid;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config.ts]: Already handled: `config.ts` line 692 already emits a `console.warn` on dimension mismatch: `memory-hybrid: embedding.models — models have mismatched vector dimensions (${dims}); all will be ignored. Models in a list must share the same output dimension.` The `${dims}` string lists every model with its dimension (e.g. `text-embedding-3-small=1536, text-embedding-3-large=3072`), so users can see exactly which models conflict.

File: extensions/memory-hybrid/services/embeddings.ts [hunk 8]
```diff
@@ -20,44 +20,65 @@ function hashText(text: string): string {
...
+    clientOrApiKey: OpenAI | string,
+    modelOrModels: string | string[],
   ) {
-    this.client = new OpenAI({ apiKey });
+    this.client = typeof clientOrApiKey === "string"
+      ? new OpenAI({ apiKey: clientOrApiKey })
+      : clientOrApiKey;
+    this.models = Array.isArray(modelOrModels) ? modelOrModels : [modelOrModels];
+    if (this.models.length === 0) throw new Error("Embeddings requires at least one model");
   }
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/embeddings.ts]: Verified: the only active call site for `new Embeddings(...)` in the live codebase is `setup/init-databases.ts:73`, which already uses the new signature `new Embeddings(openai, embeddingModels)`. The old signature `new Embeddings(apiKey, model)` appears only in `index.ts.bak` (pre-refactor backup, not compiled). No other call sites to update.

File: extensions/memory-hybrid/cli/proposals.ts [hunk 9]
```diff
@@ -341,14 +348,13 @@ export async function applyApprovedProposal(
...
-      return {
-        ok: false,
-        error: `Git commit failed; target file rolled back to original. Commit error: ${commitResult.error}`,
-      };
+    if (isGitRepo(targetPath)) {
+      const commitResult = commitProposalChange(targetPath, proposalId, proposal.targetFile);
+      if (!commitResult.ok) {
+        ctx.api?.logger?.warn?.(
+          `memory-hybrid: Git commit failed after applying proposal ${proposalId}; file was written successfully. ${commitResult.error}`,
+        );
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/cli/proposals.ts]: Fixed: Added rollback of the file write when `commitProposalChange` fails. The original content is now restored with `writeFileSync(targetPath, original)` before returning `{ ok: false, ... }`. The proposal is NOT marked as applied so the user can retry after fixing the git issue. The backup file (written before the change) is preserved for reference.

**Comment by markus-lassfolk** [comment]
## Round 3: Copilot feedback (11 threads)

### Fixed in code (5 threads)

| Thread | File | Change |
|--------|------|--------|
| Dead code after fallback loop | `services/embeddings.ts` | Removed unreachable `throw new Error("Embeddings: no model available")` — constructor guarantees ≥1 model so `lastErr` is always set after the loop |
| `?? "unused"` API key fallback | `setup/init-databases.ts` | Removed; `HybridMemoryConfig.embedding.apiKey` is typed `string` and schema throws on missing/placeholder key |
| Prompt confidence wording mismatch | `prompts/generate-proposals.txt` | `"below 0.6"` → `"0.6 maximum"` and `"0.5 or lower"` → `"0.5 maximum"` to match `Math.min` caps in `capProposalConfidence` |
| File write not rolled back on git failure | `cli/proposals.ts` | Added `writeFileSync(targetPath, original)` + early return `{ ok: false }` when `commitProposalChange` fails; proposal stays unapplied |
| `isGitRepo` directory-path false negative | `cli/proposals.ts` | Replaced trailing-slash heuristic with `statSync(path).isDirectory()` (fallback to `dirname` if stat throws) |

### Acknowledged as already handled / by design (6 threads)

- **Port validation** — `OPENCLAW_GATEWAY_PORT` already validated with `parseInt + NaN + range 1–65535` check + `warn` log (init-databases.ts:65–68)
- **Confidence cap order** — cap already applied on line 1493 *before* `minConf` check on line 1494 (correct order)
- **Unknown-model warning** — `config.ts:683` already emits `console.warn` naming the skipped model
- **Dimension-mismatch warning** — `config.ts:692` already emits `console.warn` listing all model dimensions
- **Embeddings call-site audit** — only one active call site (`init-databases.ts:73`), already on new signature
- **scopeFilter in proposal generation** — by design; prevents cross-user contamination; `warn` emitted at line 1398 when unset

---
tsc: clean | tests: 673 passed, 3 skipped (version-command.test.ts pre-existing failures ignored per instructions)
